example-events/src/ofApp.cpp

```diff
@@ -2,49 +2,50 @@
 
 //--------------------------------------------------------------
 void ofApp::setup(){
-    
     image.load("image.jpg");
     warper.setup(0,0,image.getWidth(),image.getHeight());
     warper.activate();
     
     // position the image in the center of the screen
     // screen size
-    float sw = ofGetWidth();
-    float sh = ofGetHeight();
+    sw = ofGetWidth();
+    sh = ofGetHeight();
     // image size
-    float iw = image.getWidth();
-    float ih = image.getHeight();
+    iw = image.getWidth();
+    ih = image.getHeight();
     
-    TLPosition = ofPoint(sw/2 - iw/2, sh/2 - ih/2);
-    TRPosition = ofPoint(sw/2 + iw/2, sh/2 - ih/2);
-    BLPosition = ofPoint(sw/2 - iw/2, sh/2 + ih/2);
-    BRPosition = ofPoint(sw/2 + iw/2, sh/2 + ih/2);
+    TLPosition = glm::vec2(sw/2 - iw/2, sh/2 - ih/2);
+    TRPosition = glm::vec2(sw/2 + iw/2, sh/2 - ih/2);
+    BLPosition = glm::vec2(sw/2 - iw/2, sh/2 + ih/2);
+    BRPosition = glm::vec2(sw/2 + iw/2, sh/2 + ih/2);
     
     warper.setCorner(ofxGLWarper::TOP_LEFT, TLPosition);
     warper.setCorner(ofxGLWarper::TOP_RIGHT, TRPosition);
     warper.setCorner(ofxGLWarper::BOTTOM_LEFT, BLPosition);
     warper.setCorner(ofxGLWarper::BOTTOM_RIGHT, BRPosition);
-
     //alternatively, this works the same:
     // warper.setAllCorners(TLPosition, TRPosition, BLPosition, BRPosition);
     
-    // start listening to events
-    ofAddListener(warper.changeEvent, this, &ofApp::onCornerChange);
+    // start listening to corner events, the function is only called if the value changed.
+    warper.corners[0].addListener(this, &ofApp::onCornerChange);
+    warper.corners[1].addListener(this, &ofApp::onCornerChange);
+    warper.corners[2].addListener(this, &ofApp::onCornerChange);
+    warper.corners[3].addListener(this, &ofApp::onCornerChange);
 
 }
 
 //--------------------------------------------------------------
 void ofApp::update(){
     // move the TOP_LEFT corner via code
-    ofPoint newPosition = ofPoint(ofGetWidth()/2 - image.getWidth()/2, ofGetHeight()/2 - image.getHeight()/2);
-    newPosition += ofPoint(50,50) * sin(ofGetElapsedTimef());    
+    glm::vec2 newPosition = glm::vec2(sw/2 - iw/2, sh/2 - ih/2);
+    newPosition += glm::vec2(50,50) * sin(ofGetElapsedTimef());
     warper.setCorner(ofxGLWarper::TOP_LEFT, newPosition);
     
 }
 
 //--------------------------------------------------------------
 void ofApp::draw(){
-    ofBackgroundGradient(ofColor(120,120,120), ofColor(80,80,80));
+    ofBackground(0,0,0);
```

Comment:
what's the problem with the gradient.? :)

Reply:
Oh, that was for testing, I don't remember, sorry, I'll repaint the walls just like they were before...
![](http://www.expert-peinture.fr/wp-content/uploads/2014/11/cout-des-travaux-de-peinture.jpg)